include/oneapi/dpl/pstl/algorithm_ranges_impl.h

```diff
@@ -0,0 +1,315 @@
+
+template <typename _Tag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+auto
+__pattern_for_each_impl(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+
+    auto __f_1 = [__f, __proj](auto&& __val) { std::invoke(__f, std::invoke(__proj, __val));};
+
+    oneapi::dpl::__internal::__pattern_walk1(__tag, std::forward<_ExecutionPolicy>(__exec), std::ranges::begin(__r),
+        std::ranges::begin(__r) + __r.size(), __f_1);
+
+    using __return_t = std::ranges::for_each_result<std::ranges::borrowed_iterator_t<_R>, _Fun>;
+    return __return_t{__r.begin() + __r.size(), std::move(__f)};
+}
+
+template <typename _IsVector, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+auto
+__pattern_for_each(__parallel_tag<_IsVector> __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    return __pattern_for_each_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_R>(__r), __f, __proj);
+}
+
+template <typename _Tag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+auto
+__pattern_for_each(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    if constexpr(typename _Tag::__is_vector{})
+        return __pattern_for_each_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_R>(__r), __f, __proj);
+    else
+        return std::ranges::for_each(std::forward<_R>(__r), __f, __proj);
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_transform
+//---------------------------------------------------------------------------------------------------------------------
+
+template<typename _Tag, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _F, typename _Proj>
+auto
+__pattern_transform_impl(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r,
+                    _F __op, _Proj __proj)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+    assert(__in_r.size() == __out_r.size());
+
+    auto __unary_op = [=](auto&& __val) -> decltype(auto) { return std::invoke(__op, std::invoke(__proj, __val));};
+
+    oneapi::dpl::__internal::__pattern_walk2(__tag, std::forward<_ExecutionPolicy>(__exec), std::ranges::begin(__in_r),
+        std::ranges::begin(__in_r) + __in_r.size(), std::ranges::begin(__out_r),
+        oneapi::dpl::__internal::__transform_functor<decltype(__unary_op)>{std::move(__unary_op)});
+
+    using __return_t = std::ranges::unary_transform_result<std::ranges::borrowed_iterator_t<_InRange>,
+        std::ranges::borrowed_iterator_t<_OutRange>>;
+
+    return __return_t{__in_r.begin() + __in_r.size(), __out_r.begin() + __out_r.size()};
```

Comment:
```suggestion
    return __return_t{std::ranges::begin(__in_r) + __in_r.size(), std::ranges::begin(__out_r) + __out_r.size()};
```

Reply:
yeah, `std::ranges::begin` is better. I think it should be changed in others similar places as well.
Thank you, Adam.